### ISSUE TITLE
Clasificador de redes Convulsionados

### ISSUE THREAD

**Issue by HeladeroTV** [HeladeroTV/Tareas-IA#6]
``import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.callbacks import EarlyStopping

# Configuración de semilla para reproducibilidad
np.random.seed(42)
tf.random.set_seed(42)

def crear_modelo_cnn(input_shape=(28, 28, 1), num_clases=10):
    """
    Crea un modelo CNN simple para clasificación de imágenes
    
    Args:
        input_shape: Forma de la imagen de entrada (altura, ancho, canales)
        num_clases: Número de clases para clasificar
        
    Returns:
        Modelo CNN compilado
    """
    modelo = Sequential([
        # Primera capa convolucional
        Conv2D(32, kernel_size=(3, 3), activation='relu', padding='same', input_shape=input_shape),
        MaxPooling2D(pool_size=(2, 2)),
        
        # Segunda capa convolucional
        Conv2D(64, kernel_size=(3, 3), activation='relu', padding='same'),
        MaxPooling2D(pool_size=(2, 2)),
        
        # Tercera capa convolucional
        Conv2D(128, kernel_size=(3, 3), activation='relu', padding='same'),
        MaxPooling2D(pool_size=(2, 2)),
        
        # Aplanar para las capas densas
        Flatten(),
        
        # Capas densas
        Dense(128, activation='relu'),
        Dropout(0.5),  # Regularización para evitar sobreajuste
        Dense(num_clases, activation='softmax')  # Capa de salida
    ])
    
    # Compilar el modelo
    modelo.compile(
        optimizer='adam',
        loss='categorical_crossentropy',
        metrics=['accuracy']
    )
    
    return modelo

def entrenar_modelo(modelo, x_train, y_train, x_val, y_val, batch_size=32, epocas=20):
    """
    Entrena el modelo CNN con aumento de datos
    
    Args:
        modelo: Modelo CNN compilado
        x_train, y_train: Datos de entrenamiento y etiquetas
        x_val, y_val: Datos de validación y etiquetas
        batch_size: Tamaño del lote
        epocas: Número máximo de épocas de entrenamiento
        
    Returns:
        Historia del entrenamiento
    """
    # Crear generadores de datos con aumento para mejorar la generalización
    datagen = ImageDataGenerator(
        rotation_range=10,
        width_shift_range=0.1,
        height_shift_range=0.1,
        zoom_range=0.1,
        horizontal_flip=True
    )
    
    # Configurar early stopping para detener el entrenamiento cuando no hay mejora
    early_stopping = EarlyStopping(
        monitor='val_accuracy',
        patience=5,
        restore_best_weights=True
    )
    
    # Entrenar el modelo
    historia = modelo.fit(
        datagen.flow(x_train, y_train, batch_size=batch_size),
        steps_per_epoch=len(x_train) // batch_size,
        epochs=epocas,
        validation_data=(x_val, y_val),
        callbacks=[early_stopping]
    )
    
    return historia

def evaluar_modelo(modelo, x_test, y_test):
    """
    Evalúa el modelo en los datos de prueba
    
    Args:
        modelo: Modelo CNN entrenado
        x_test, y_test: Datos y etiquetas de prueba
        
    Returns:
        Pérdida y precisión en los datos de prueba
    """
    perdida, precision = modelo.evaluate(x_test, y_test, verbose=0)
    print(f"Precisión en datos de prueba: {precision * 100:.2f}%")
    return perdida, precision

# Ejemplo de uso con MNIST (un conjunto de datos estándar para comenzar)
if __name__ == "__main__":
    # Cargar el conjunto de datos MNIST
    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
    
    # Preprocesar datos
    # Normalizar valores de píxeles al rango [0, 1]
    x_train = x_train.astype('float32') / 255
    x_test = x_test.astype('float32') / 255
    
    # Añadir dimensión de canal (MNIST es de un solo canal - escala de grises)
    x_train = np.expand_dims(x_train, -1)
    x_test = np.expand_dims(x_test, -1)
    
    # Convertir etiquetas a one-hot encoding
    y_train = tf.keras.utils.to_categorical(y_train, 10)
    y_test = tf.keras.utils.to_categorical(y_test, 10)
    
    # Separar datos de validación
    val_size = 5000
    x_val = x_train[-val_size:]
    y_val = y_train[-val_size:]
    x_train = x_train[:-val_size]
    y_train = y_train[:-val_size]
    
    # Crear y compilar el modelo
    modelo = crear_modelo_cnn(input_shape=(28, 28, 1), num_clases=10)
    modelo.summary()
    
    # Entrenar el modelo
    historia = entrenar_modelo(modelo, x_train, y_train, x_val, y_val)
    
    # Evaluar el modelo
    evaluar_modelo(modelo, x_test, y_test)
    
    # Guardar el modelo entrenado
    modelo.save('modelo_cnn_simple.h5')
    print("Modelo guardado como 'modelo_cnn_simple.h5'")``

Un Clasificador simple el cual utiliza el DATASET MNIST de digitos escritos a mano, cuenta con una arquitectura simple de 3 capas convolucionales
Con este podemos crear, entrenar y evaluar el modelo y para ejecutar no necesitamos meter cosas externas
